test/performance/io/format_fastq_benchmark.cpp

```diff
@@ -0,0 +1,285 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+// @author: Clemens Cords <clemens.cords@fu-berlin.de>
+
+#include <benchmark/benchmark.h>
+#include <cstring>
+#include <fstream>
+
+#define SEQAN2_HAS_SEQAN3 __has_include(<seqan/seq_io.h>)
+#if SEQAN2_HAS_SEQAN3
+    #include <seqan/seq_io.h>
+#endif
+
+#include <seqan3/alphabet/nucleotide/dna5.hpp>
+#include <seqan3/alphabet/quality/phred42.hpp>
+#include <seqan3/io/sequence_file/format_fastq.hpp>
+#include <seqan3/io/sequence_file/input.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/test/performance/sequence_generator.hpp>
+#include <seqan3/std/filesystem>
+
+using namespace seqan3;
+
+unsigned int const SEED = 1234;
+
+// ============================================================================
+// generate fastq file
+// ============================================================================
+constexpr std::size_t SEQUENCE_LENGTH = 50; //length of nucleotide and quality sequence
+constexpr std::size_t N_ENTRIES_IN_FILE = 4069; // default number of 3-line entries
+
+// if a file of length m is requested once, that file is stored so
+// any functions later on that needs the exact same file can access it
+std::map<std::size_t, std::string> file_dict;
+
+std::string get_file(std::size_t n_entries)
```

Comment:
I"m not sure if this is the way to go. I understand your intention to ensure that every benchmark has the same input, so that the benchmark results are [more] comparable to each other AND to be efficient and don't recompute the same. But, I think in this case it would suffice to write a `generate_fasta_file` function without any memorization/result-caching.

Can you put everything within the `if (!has_been_init)` into a own function `generate_fasta_file`?

Reply:
Do the contents of the file (what permutation of nucleotides / phred-scores) have absolutely no impact on runtime? If so, I'll edit it to just generate a new file every time since in that case you're absolutely right and as long as the files have the same number of chars the benchmarks are comparable.

My rational behind caching it was that I figured at some point during the parsing of the file there will be a moment where the current char c is compared against each introducing a constant depending on the char c like so:
```cpp
// pseudocode
nucleotide c = (...)
switch (c)
{
    case A:
        (...)
        break;

    case C:
        (...)
        break;

    case G:
        (...)
        break;

    case T:
        (...)
        break;
} 
```
Then if ``c == A`` it will introduce a constant of 1 (1 if-comparison), while ``c == T`` will introduce a constant of 4 (4 if-comparisons). For bigger alphabets like phred42 this means the variance can be up to 41 operations per char which would mean for a file with 10^6 lines this could mean at worst a difference of 41 *10^6 operations.

I know being this stingy about runtime in something like a benchmark that isn't really important (and has crazy variance just based on the cpu workload during execution) isn't necessary but I don't see a reason to actively make something less exact.

Please let me know, this is more about me learning how it works and out of principle than how the benchmark turns out. I already put everything in it's own function but kept the memory caching for now

Reply:
If the value of the characters have an influence on the run time, we should notice some jitter in the results.
----

In this case the compiler is smarter than you :)

`A,C,G,T` are basically a fancy way of writing the numbers `0, 1, 2, 3`.

See  https://godbolt.org/z/M26FVb for an example what the compiler does in those cases.

Spoiler: this is the complete generated assembly; independent on how many cases you have. Tip: try to add the more cases, the assembly will stay at 5 instructions.
```
        mov     eax, DWORD PTR [rsp+12]
        xor     r8d, r8d
        cmp     eax, 3
        ja      .L1
        mov     r8d, DWORD PTR CSWTCH.3[0+rax*4]
```

Reply:
I removed the caching. This caused runtime to quadruple (since a new file has to be generated from scratch each template call)